parquet-hadoop/src/test/java/org/apache/parquet/hadoop/TestZstandardCodec.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.parquet.example.data.simple.SimpleGroupFactory;
+import org.apache.parquet.hadoop.codec.ZstandardCodec;
+import org.apache.parquet.hadoop.example.GroupWriteSupport;
+import org.apache.parquet.hadoop.mapred.DeprecatedParquetOutputFormat;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+import org.apache.parquet.schema.MessageTypeParser;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.util.Random;
+
+public class TestZstandardCodec {
+
+  private final Path inputPath = new Path("src/test/java/org/apache/parquet/hadoop/example/TestInputOutputFormat.java");
+
+  @Test
+  public void testZstdCodec() throws IOException {
+    ZstandardCodec codec = new ZstandardCodec();
+    Configuration conf = new Configuration();
+    int[] levels = {1, 4, 7, 10, 13, 16, 19, 22};
+    int[] dataSizes = {0, 1, 10, 1024, 1024 * 1024};
+
+    for (int i = 0; i < levels.length; i++) {
+      conf.setInt(ZstandardCodec.PARQUET_COMPRESS_ZSTD_LEVEL, levels[i]);
+      codec.setConf(conf);
+      for (int j = 0; j < dataSizes.length; j++) {
+        testZstd(codec, dataSizes[j]);
+      }
+    }
+  }
+
+  private void testZstd(ZstandardCodec codec, int dataSize) throws IOException {
+    byte[] data = new byte[dataSize];
+    (new Random()).nextBytes(data);
+    BytesInput compressedData = compress(codec,  BytesInput.from(data));
+    BytesInput decompressedData = decompress(codec, compressedData, data.length);
+    Assert.assertArrayEquals(data, decompressedData.toByteArray());
+  }
+
+  private BytesInput compress(ZstandardCodec codec, BytesInput bytes) throws IOException {
+    ByteArrayOutputStream compressedOutBuffer = new ByteArrayOutputStream((int)bytes.size());
+    CompressionOutputStream cos = codec.createOutputStream(compressedOutBuffer, null);
+    bytes.writeAllTo(cos);
+    cos.close();
+    return BytesInput.from(compressedOutBuffer);
+  }
+
+  private BytesInput decompress(ZstandardCodec codec, BytesInput bytes, int uncompressedSize) throws IOException {
+    BytesInput decompressed;
+    InputStream is = codec.createInputStream(bytes.toInputStream(), null);
+    decompressed = BytesInput.from(is, uncompressedSize);
+    is.close();
+    return decompressed;
+  }
+
+  @Test
+  public void testZstdConfWithMr() throws Exception {
+    JobConf jobConf = new JobConf();
+    Configuration conf = new Configuration();
+    jobConf.setInt(ZstandardCodec.PARQUET_COMPRESS_ZSTD_LEVEL, 18);
+    jobConf.setInt(ZstandardCodec.PARQUET_COMPRESS_ZSTD_LEVEL, 4);
+    RunningJob mapRedJob = runMapReduceJob(CompressionCodecName.ZSTD, jobConf, conf);
+    assert(mapRedJob.isSuccessful());
+  }
+
+  private RunningJob runMapReduceJob(CompressionCodecName codec, JobConf jobConf, Configuration conf) throws IOException, ClassNotFoundException, InterruptedException {
```

Comment:
Thanks for creating an MR test. What I wanted to test with it is if the ZSTD related configuration takes place if set via the configuration. This is not tested currently. I wanted to test it because I am not sure if we have the code in place to pass configuration properties to the codecs.

Reply:
I set breakpoint and checked it. But I don't know how to get back the level/workers at the MR job. Any ideal how to do that? I also verified it in the standalone Spark application. Do you think that is sufficient?

Reply:
Maybe, if we create two parquet files with different levels (e.g. 1 and 22) but with the exact same data we can expect that the larger level will generate smaller file?
I just want to ensure that the properties really arrive to the codec. If you think the related test requires too much effort or even not feasible I am fine if you can point me to the code path where the properties are passed through to the codecs.

Reply:
I like the idea to compare the two files with different compression levels. Only thing is that it is not 100% guarantee the higher level always gets higher compression ratio although the chance of that is very low. In our test data, it seems it is OK though. So I will just go ahead to add the file size comparison.

Reply:
Yes, I know, there is no 100% guarantee in normal circumstances. But if you write the test in a deterministic way (as it should be by e.g. using a fix seed for random generation) then without changing the data and the compression codec it shall be fine. I worth a comment, though about it is not the nicest way of testing the work of the properties but currently that's seems to be the only one.

Reply:
Thanks.